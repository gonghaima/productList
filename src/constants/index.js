import keyMirror from 'fbjs/lib/keyMirror';

/**
 * @namespace Constants
 * @desc App constants
 */

/**
 * @constant {Object} ActionTypes
 * @memberof Constants
 */
export const ActionTypes = keyMirror({
  SWITCH_MENU: undefined,
  EXCEPTION: undefined,
  PRODUCT_GET_REPOS: undefined,
  PRODUCT_GET_REPOS_SUCCESS: undefined,
  PRODUCT_GET_REPOS_FAILURE: undefined,
  SELECTION_CHANGE: undefined,
  PAGE_CHANGE: undefined,
});

/**
 * @constant {Object} STATUS
 * @memberof Constants
 */
export const STATUS = {
  IDLE: 'idle',
  RUNNING: 'running',
  READY: 'ready',
  SUCCESS: 'success',
  ERROR: 'error',
};
