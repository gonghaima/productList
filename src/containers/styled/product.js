import styled from 'styled-components';
import { titleColor, borderColor, itemBgColor, subTitleColor, focusedColor } from 'modules/theme';

import { Box, utils } from 'styled-minimal';

const { responsive, spacer } = utils;

const Title = styled.h2`
  color: ${titleColor};
  padding: 0 10px;
  margin: 0;
`;

const Summary = styled.div`
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  justify-content: space-between;
  padding: 0 10px;
  color: ${subTitleColor};
`;
const SummaryItem = styled.div`
  margin: 0;
  width: 10%;
  select {
    background-color: #f6f6f6 !important;
  }
`;

const ProductCount = styled.span`
  vertical-align: sub;
`;

const Divider = styled.hr`
  display: block;
  height: 1px;
  border: 0;
  border-top: 1px solid ${borderColor};
  margin: 0.7em 0.4em;
  padding: 0;
`;
const ProductGrid = styled.ul`
  display: grid;
  grid-auto-flow: row;
  grid-gap: ${spacer(2)};
  grid-template-columns: 100%;
  list-style: none;
  margin: 0px ${spacer(4)} auto 0;
  padding: 0;
  /* stylelint-disable */
  ${/* istanbul ignore next */ p =>
    responsive({
      ix: `
        grid-gap: ${spacer(3)(p)};
        width: 90%;
      `,
      md: `
        grid-template-columns: repeat(2, 1fr);
        width: 100%;
      `,
      lg: `
        grid-template-columns: repeat(3, 1fr);
      `,
      xl: `
        grid-gap: ${spacer(4)(p)};
        grid-template-columns: repeat(4, 1fr);
      `,
    })};
  /* stylelint-enable */

  > li {
    display: flex;
  }
`;

const Item = styled(Box)`
  background: ${itemBgColor};
  align-items: center;
  border: solid 0.1rem ${borderColor};
  border-radius: 0.4rem;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  overflow: hidden;
  padding: ${spacer(3)};
  text-align: center;
  width: 100%;
  /* stylelint-disable */
  ${/* istanbul ignore next */ p =>
    responsive({
      md: `
        padding: ${spacer(3)(p)} ${spacer(0)(p)};
      `,
      lg: `
        padding: ${spacer(2)(p)} ${spacer(0)(p)};
      `,
    })};
  /* stylelint-enable */

  > a {
    margin-bottom: ${spacer(2)};
  }

  img {
    height: 20rem;
  }
`;

const ImageSection = styled.div`
  padding: 6px;
  margin: 0;
`;
const DetailsSection = styled.div`
  border-top: 1px solid ${borderColor};
  text-align: left;
  padding-left: ${spacer(3)};
  width: 100%;
  display: block;
  margin-bottom: ${spacer(3)};

  a {
    display: block;
  }
`;

const ItemTitle = styled.h4`
  color: ${titleColor};
  margin: ${spacer(3)} ${spacer(0)} ${spacer(1)} ${spacer(0)};
`;

const ItemDescription = styled.h4`
  color: ${subTitleColor};
  margin: 0px 0px ${spacer(2)} 0px;
`;

const ItemPrice = styled.h6`
  color: ${focusedColor};
  margin: 0;
`;

const Pagination = styled.div`
  display: flex;
  flex-direction: row-reverse;
  ul {
    display: inline-block;
    padding-left: 15px;
    padding-right: 15px;
  }

  li {
    display: inline-block;
  }
  a {
    z-index: 3;
    color: #000;
    cursor: pointer;
    position: relative;
    float: left;
    padding: 6px 12px;
    margin-left: -1px;
    line-height: 1.42857143;
    text-decoration: none;
  }
  .active a:hover,
  .active a {
    background: #fff;
    outline: unset;
    border-bottom: 3px solid ${focusedColor};
  }
  .disabled a {
    color: ${subTitleColor};
    cursor: not-allowed;
  }
`;

export {
  Title,
  Summary,
  SummaryItem,
  ProductCount,
  Divider,
  ProductGrid,
  Item,
  ImageSection,
  DetailsSection,
  ItemTitle,
  ItemDescription,
  ItemPrice,
  Pagination,
};
