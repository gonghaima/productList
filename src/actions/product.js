// @flow
/**
 * @module Actions/User
 * @desc User Actions
 */
import { createActions } from 'redux-actions';

import { ActionTypes } from 'constants/index';

export const { productGetMake: getProducts } = createActions({
  [ActionTypes.PRODUCT_GET_MAKE]: query => ({ query }),
});
