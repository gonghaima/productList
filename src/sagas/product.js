/**
 * @module Sagas/Product
 * @desc Product
 */

import { all, call, put, takeLatest } from 'redux-saga/effects';
import { request } from 'modules/client';
import { ActionTypes } from 'constants/index';

/**
 * Get Repos
 *
 * @param {Object} action
 *
 */
export function* getProducts() {
  try {
    const response = yield call(request, '/api/product');
    yield put({
      type: ActionTypes.PRODUCT_GET_SUCCESS,
      payload: { data: (response && response.data) || [] },
    });
  } catch (err) {
    yield put({
      type: ActionTypes.PRODUCT_GET_REPOS_FAILURE,
      payload: err,
    });
  }
}

/**
 * Product Sagas
 */
export default function* root() {
  yield all([takeLatest(ActionTypes.PRODUCT_GET_MAKE, getProducts)]);
}
