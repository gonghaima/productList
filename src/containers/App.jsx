import React from 'react';
// import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Router, Switch, Route } from 'react-router-dom';
import Helmet from 'react-helmet';
import styled, { ThemeProvider } from 'styled-components';

import history from 'modules/history';
import theme, { appColor, titleColor } from 'modules/theme';

import config from 'config';

import Home from 'routes/Home';
import NotFound from 'routes/NotFound';

import GlobalStyles from 'components/GlobalStyles';
import RouteProduct from 'components/RouteProduct';

const AppWrapper = styled.div`
  display: flex;
  background-color: ${appColor};
  flex-direction: column;
  min-height: 100vh;
  opacity: 1 !important;
  position: relative;
  transition: opacity 0.5s;
  h3,
  h5 {
    color: ${titleColor};
  }
`;

const Main = styled.main`
  min-height: 100vh;
`;

export class App extends React.Component {
  static propTypes = {};

  render() {
    return (
      <Router history={history}>
        <ThemeProvider theme={theme}>
          <AppWrapper>
            <Helmet
              defer={false}
              htmlAttributes={{ lang: 'pt-br' }}
              encodeSpecialCharacters={true}
              defaultTitle={config.title}
              titleTemplate={`%s | ${config.name}`}
              titleAttributes={{ itemprop: 'name', lang: 'pt-br' }}
            />
            <Main>
              <Switch>
                <RouteProduct path="/" component={Home} />
                <Route component={NotFound} />
              </Switch>
            </Main>
            <GlobalStyles />
          </AppWrapper>
        </ThemeProvider>
      </Router>
    );
  }
}

export default connect()(App);
