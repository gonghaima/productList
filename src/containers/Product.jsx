import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import ReactPaginate from 'react-paginate';
import { connect } from 'react-redux';

import { getRepos, setSelection, setPage } from 'actions/index';
import { STATUS } from 'constants/index';

import { Select, Flex, Image } from 'styled-minimal';
import Loader from 'components/Loader';
import {
  Title,
  Summary,
  SummaryItem,
  ProductCount,
  Divider,
  ProductGrid,
  Item,
  ImageSection,
  DetailsSection,
  ItemTitle,
  ItemDescription,
  ItemPrice,
  Pagination,
} from './styled/product';

export class Product extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      query: 'react',
    };

    this.handleSelect = this.handleSelect.bind(this);
    this.handlePageChange = this.handlePageChange.bind(this);
  }

  static propTypes = {
    dispatch: PropTypes.func.isRequired,
    product: PropTypes.object.isRequired,
    selection: PropTypes.object.isRequired,
  };

  componentDidMount() {
    const { query } = this.state;
    const { dispatch } = this.props;
    dispatch(getRepos(query));
  }

  handleSelect(event) {
    const { dispatch } = this.props;
    dispatch(setSelection(event.target.value));
  }

  handlePageChange(event) {
    // {selected: 1}

    const {
      dispatch,
      selection: { itemPerPage },
    } = this.props;
    const offset = Math.floor(event.selected * itemPerPage);
    dispatch(setPage(offset));
  }

  render() {
    const { query } = this.state;
    const { product, selection } = this.props;
    const data = product.repos.data[query] || [];
    const pageCount =
      selection && selection.itemPerPage ? Math.ceil(data.length / selection.itemPerPage) : 0;
    let output;
    if (product.repos.status === STATUS.READY) {
      if (data.length) {
        output = (
          <Fragment>
            <Title>All Products</Title>
            <Summary>
              <SummaryItem>
                <ProductCount>{data.length} Products</ProductCount>
              </SummaryItem>

              <SummaryItem>
                <Select sizing="sm" bordered={false} onChange={this.handleSelect}>
                  {selection.selectionItems.map(val => (
                    <option key={val} value={val}>
                      {val} per page
                    </option>
                  ))}
                </Select>
              </SummaryItem>
            </Summary>
            <Divider />
            selection.currentPage{selection.currentPage}***** selection.itemPerPage{' '}
            {selection.itemPerPage}***** selection.currentPage * selection.itemPerPage{' '}
            {selection.currentPage * selection.itemPerPage}
            <ProductGrid data-type={query} data-testid="ProductGrid">
              {product.repos.data[query]
                .slice(selection.currentPage, selection.itemPerPage)
                .map(d => (
                  <li key={d.id}>
                    <Item>
                      <ImageSection>
                        <Image src={d.product_image} alt="alt" />
                      </ImageSection>
                      <DetailsSection>
                        <ItemTitle>{d.product_name}</ItemTitle>
                        <ItemDescription>{d.description}</ItemDescription>
                        <ItemPrice>{d.price}</ItemPrice>
                      </DetailsSection>
                    </Item>
                  </li>
                ))}
            </ProductGrid>
            <Pagination>
              <ReactPaginate
                previousLabel={'< previous'}
                nextLabel={'next >'}
                breakLabel="..."
                breakClassName="break-me"
                pageCount={pageCount}
                marginPagesDisplayed={2}
                pageRangeDisplayed={5}
                onPageChange={this.handlePageChange}
                containerClassName="pagination"
                subContainerClassName="pages pagination"
                activeClassName="active"
              />
            </Pagination>
          </Fragment>
        );
      } else {
        output = <h3>Nothing found</h3>;
      }
    } else {
      output = <Loader block />;
    }

    return (
      <div key="Product" data-testid="ProductWrapper">
        <Flex justifyContent="center" />
        {output}
      </div>
    );
  }
}

/* istanbul ignore next */
function mapStateToProps(state) {
  return { product: state.product, selection: state.selection };
}

export default connect(mapStateToProps)(Product);
