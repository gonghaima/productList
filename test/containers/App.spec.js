import React from 'react';

import { STATUS } from 'constants/index';

import { App } from 'containers/App';
import { Pagination } from '../../src/containers/styled/product';

jest.mock('uuid/v4', () => () => 'ABCDE');

const mockDispatch = jest.fn();
const props = {
  dispatch: mockDispatch,
  product: {
    repos: {
      data: [],
    },
  },
};

function setup(ownProps = props) {
  return shallow(<App {...ownProps} />, { attachTo: document.getElementById('react') });
}

describe('App', () => {
  const wrapper = setup();

  it('should render properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should not render if selected data doesn't exist", () => {
    wrapper.setProps({
      product: {},
    });

    expect(wrapper.find('Grid')).not.toExist();
  });

  it('should render the Grid if data exists', () => {
    wrapper.setProps({
      selection: {
        itemPerPage: 1,
        selectionItems: [1, 3, 5],
      },
      product: {
        repos: {
          data: {
            react: [
              {
                id: 12,
                price: '$58.30',
                product_name: 'CAPTOPRIL',
                description: 'synthesize seamless methodologies',
                product_image: 'http://dummyimage.com/340x330.jpg/cc0000/ffffff',
              },
            ],
          },
          status: STATUS.READY,
        },
      },
    });

    expect(wrapper.find('Grid')).toMatchSnapshot();
    expect(wrapper.find(Pagination)).toMatchSnapshot();
  });
});
