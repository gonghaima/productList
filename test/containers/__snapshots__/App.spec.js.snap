// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render properly 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "query": Object {},
        "search": "",
        "state": Object {},
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <ThemeProvider
    theme={
      Object {
        "breakpoints": Array [
          400,
          768,
          1024,
          1280,
          1920,
        ],
        "button": Object {
          "borderRadius": Object {
            "lg": 28,
            "xl": 32,
            "xs": 4,
          },
          "padding": Object {
            "lg": Array [
              12,
              28,
            ],
            "xl": Array [
              14,
              32,
            ],
          },
        },
        "palette": Object {
          "anchorHoverColor": "#337ab7",
          "borderColor": "#E5E5E5",
          "mainBgColor": "#f6f8fa",
          "primary": "#f6f6f6",
        },
      }
    }
  >
    <AppWrapper>
      <HelmetWrapper
        defaultTitle="react-redux-saga-boilerplate"
        defer={false}
        encodeSpecialCharacters={true}
        htmlAttributes={
          Object {
            "lang": "pt-br",
          }
        }
        titleAttributes={
          Object {
            "itemprop": "name",
            "lang": "pt-br",
          }
        }
        titleTemplate="%s | react-redux-saga-boilerplate"
      />
      <Main>
        <Switch>
          <RouteProduct
            component={[Function]}
            path="/"
          />
          <Route
            component={[Function]}
          />
        </Switch>
      </Main>
      <_default />
    </AppWrapper>
  </ThemeProvider>
</Router>
`;

exports[`App should render the Grid if data exists 1`] = `null`;

exports[`App should render the Grid if data exists 2`] = `null`;
